### ISSUE TITLE
Feature Request - Ability to terminate the node at the specified time

### ISSUE THREAD

**Issue by vikas027** [aws/karpenter#1302]
**Tell us about your request**
We already have a way to terminate the instance after X amount of time. It would be great to have a way of doing the same at a specified time as well. For example, say I want to terminate a node(s) at 2:00 AM UTC to save running costs.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
We are looking to have the ability to scale down nodes at certain times, with ASGs it is easy using the scheduled actions but since Karpenter is managed outside of an ASG (which is a good thing), we can't automatically scale them down to save costs.

**Are you currently working around this issue?**
I may end up writing a lambda function to do the same but it would be great to have this feature inbuilt

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by bwagner5** [comment]
This is an interesting request, thanks for posting! We have considered adding a maintenance window to the node expiration, so that if a node expires, it will wait to terminate the node until the maintenance window. We haven't committed to building this feature but it's a similar idea although solving different end-goals. 

I suppose your use-case is to spin down dev workloads or something that is not being used for cost?  I don't think this use-case is something that Karpenter should necessarily deal with directly. But Karpenter would enable something like this to be built really easily with a Kubernetes cron-job that simply `kubectl delete node <nodes>` or to `kubectl scale deploy/<deployment> --replicas=0` which can scale down the non-critical workloads and if the nodes are empty, Karpenter will automatically delete them.  A lambda would be super easy to build as well, as you suggested.

**Comment by vikas027** [comment]
>  I don't think this use-case is something that Karpenter should necessarily deal with directly. 

Yeah, I agree. Just raised this request as we already have a TTL feature :)

**Comment by albertjunkim** [comment]
Thanks for raising this issue, @vikas027. 

I want to add another use case of having a maintenance windows feature. 
Cost-saving for dev workloads is for sure but having a specific window allows more control and resiliency when nodes need to be updated(new patched AMI...etc.)
We wouldn't want to bounce the nodes while there are incoming live traffics, especially in production env. 
We would want to bounce the nodes when the traffic is low(e.g. during off-hours or weekends), 
This would be much low risk and low impact. 

I have this dev work done already and running tests.
I will soon raise a PR. 
Meanwhile, here is the preview of `maintenanceWindows` in provisioner CRD.
Note that `maintenanceWindows` will be active when `ttlSecondsUntilExpired` is passed.

```
labels:
  ...
providers: 
  ...
maintenanceWindows
  - weekDays: ["SAT","SUN"]
    startTime: "00:00"
    timeZone: "America/New_York"
    duration: "24:00"
```

You may specify more than one window. (e.g., off-hours on weekdays + whole weekends) 
I will cover the designs more in-depth in PR.

**Comment by ellistarn** [comment]
We've talked about this idea a fair bit in working group. I think it would pair really well with NodeDisruptionBudgets

**Comment by njtran** [comment]
Hey all, if you're tracking this, come give some feedback on https://github.com/aws/karpenter-core/pull/207!

**Comment by ceastman-r7** [comment]
I have karpenter setup with drift detection enabled but it would be good to not have karpenter based nodes replaced as soon as a new ami is available. It would be preferable to have the option to have the nodes replaced on a cron schedule. For instance each weekday at 9am central see if any nodes have drifted and subsequently need to be replaced.

**Comment by mballoni** [comment]
I got to this thread thanks to @tzneal .
 My use case is for those nodes intended for nginx controller.
 Each pod runs on a very small dedicated node. These nodes are added to the Load Balancer Target group by aws lb controller.
 However there is a connection TTL between the Load Balancer and the nodes (usually around 120 seconds) and when a node dies it may take up to that ttl where some requests will be forwarded and then fail.
  The maintenance window may dramatically reduce these requests error rate if configured to a quieter time of the day when ttlSecondsUntilExpired takes over.

**Comment by sidewinder12s** [comment]
> I got to this thread thanks to @tzneal . My use case is for those nodes intended for nginx controller. Each pod runs on a very small dedicated node. These nodes are added to the Load Balancer Target group by aws lb controller. However there is a connection TTL between the Load Balancer and the nodes (usually around 120 seconds) and when a node dies it may take up to that ttl where some requests will be forwarded and then fail. The maintenance window may dramatically reduce these requests error rate if configured to a quieter time of the day when ttlSecondsUntilExpired takes over.

I think the AWS Load Balancer Controller team recommends Pod readiness gates to reduce issues with integrating LBs and Pods in AWS.

https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.5/deploy/pod_readiness_gate/

**Comment by sjmiller609** [comment]
> > I got to this thread thanks to @tzneal . My use case is for those nodes intended for nginx controller. Each pod runs on a very small dedicated node. These nodes are added to the Load Balancer Target group by aws lb controller. However there is a connection TTL between the Load Balancer and the nodes (usually around 120 seconds) and when a node dies it may take up to that ttl where some requests will be forwarded and then fail. The maintenance window may dramatically reduce these requests error rate if configured to a quieter time of the day when ttlSecondsUntilExpired takes over.
> 
> I think the AWS Load Balancer Controller team recommends Pod readiness gates to reduce issues with integrating LBs and Pods in AWS.
> 
> https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.5/deploy/pod_readiness_gate/

Pod readiness works great for me in startup, and pod pre-stop hook and termination grace period are necessary to prevent the pod from being removed before the load balancer is done draining connections (for me on AWS NLB can take up to 180 seconds to finish draining connections)

```
            # These configurations allow for the NLB to drain connections
            # An NLB can take up to 180 seconds to stop routing TCP connections
            # to a target.
            terminationGracePeriodSeconds: 200
            lifecycle:
              preStop:
                exec:
                  command: ["/bin/sh", "-c", "sleep 180"]
```

This pairs well with surged deployment to reduce how long it takes

```
          updateStrategy:
            type: RollingUpdate
            rollingUpdate:
              maxUnavailable: 0
              maxSurge: 100%
```

**Comment by sjmiller609** [comment]
> Hey all, if you're tracking this, come give some feedback on [aws/karpenter-core#207](https://github.com/aws/karpenter-core/pull/207)!

@njtran This issue is now closed, is the next step for RFCs that are for each use case? Our team in interested in the maintenance windows concept because our workload benefits from less frequent pod restarts.

For others in the same situation, in the meantime we are going to try the cronjob approach: https://github.com/aws/karpenter/issues/2308#issuecomment-1223044096

Also appreciate any reader if they have example of this pattern we can start from.

**Comment by portswigger-tim** [comment]
It sounds like you want: https://codeberg.org/hjacobs/kube-downscaler

**Comment by alexvaque** [comment]
Hello, we were checking the amount of releases of the AMI used on EKS, for example the ubuntu-focal-20.04-amd64-server  and we realized that the releases can be done on friday or weekends as well. So this new feature that you are proposing here is a very interesting option to avoid to rolling out K8S nodes on weekends. Thank you very much.

**Comment by njtran** [comment]
Hey all, I've linked an https://github.com/aws/karpenter-core/pull/516 that has some of the API decisions we're thinking about for some of the linked items in this issue. Please review if you get a chance!
